File: src/leetcode/0876 Middle of the Linked List/solution.js
import ListNode from '../../data_structures/LinkedListNode.js'

/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
function middleNode(head) {
    let fast = head
    let slow = head

    while (fast && fast.next) {
        fast = fast.next.next
        slow = slow.next
    }
    
    return slow
}

// Tests
function makeLinkedList(arr) {
    let currNode = new ListNode(arr[0])
    const head = currNode

    for (let i = 1; i < arr.length; i++) {
        currNode.next = new ListNode(arr[i])
        currNode = currNode.next
    }

    return head
}

console.log(middleNode(makeLinkedList([1,2,3,4,5])), 'should be 3')
console.log(middleNode(makeLinkedList([1,2,3,4,5,6])), 'should be 4')

File: src/leetcode/0876 Middle of the Linked List/solution.rs
// Source : https://leetcode.com/problems/middle-of-the-linked-list
// Author : Frank Faustino
// Date   : 2024-03-07

// Definition for singly-linked list.
#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
  pub val: i32,
  pub next: Option<Box<ListNode>>
}

impl ListNode {
  #[inline]
  fn new(val: i32) -> Self {
    ListNode {
      next: None,
      val
    }
  }
}

struct Solution;

impl Solution {
    pub fn middle_node(head: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        let mut fast = &head;
        let mut slow = &head;

        while fast.is_some() && fast.as_ref().unwrap().next.is_some() {
            fast = &fast.as_ref().unwrap().next.as_ref().unwrap().next;
            slow = &slow.as_ref().unwrap().next;
        }

        slow.clone()
    }
}

